src/core/pratt.ml

```diff
@@ -0,0 +1,113 @@
+(** This module defines functions needed by the Pratt parser of the Pratter
+    library. *)
+
+open Syntax
+
+module Pratt : sig
+  val parse : Sig_state.t -> Env.t -> p_term -> p_term
+end = struct
+
+  open Lplib
+  open Extra
+  open Pos
+
+  (** [find_qid ss env qid] fetches the qualified identifier associated to
+      [qid] if [qid] is the identifier of an infix operator defined in
+      signature state [ss]. If [qid] is defined in environment [env], then it
+      designates a variable, and is thus returned as-is. *)
+  (* REVIEW: this function relies on a lot of invariants and performs many
+     [Map.find_opt]s. Things may be optimised by modifying the data structure
+     of the signature state, notably making the qualified identfier to which
+     operators are linked available in the [unops] field of the signature
+     state. *)
+  let find_qid : Sig_state.t -> Env.t -> qident -> qident =
+    fun ss env ({elt=(mp, s); _} as qid) ->
+    try
+      if mp <> [] then raise Not_found;
+      ignore (Env.find s env);
+      qid
+    with Not_found ->
```

Comment:
Here, you know that it cannot be a variable. So it must be a symbol. You can then call Sig_state.find_sym to properly know which sym it actually is. Then, to know whether it is unary or binary, you can look into pp_hint.

Reply:
By the way, we may perhaps get rid of unops and binops as these maps have been introduced for Earley? pp_hints seems to be enough.

Reply:
Indeed, the sig state may be lightened

Reply:
In its current form, `Sig_state.find_sym` doesn't fetch the qid of operators, if I'm not mistaken, that's why I had to create this function.

Reply:
Right, it returns a sym, but you can build a qualified ident from a sym.

Reply:
Btw, it seems to me that many things could simplified in Sign.t and Sig_state.t, but perhaps in another PR. First, sign_idents could be removed. Then, sign_symbols, sign_unops, sign_binops and sign_quants could be merged into a single sign_symbols:(sym*popt*pp_hint)StrMap.t ref if notations only comes with symbol declarations. Otherwise, if notations can be defined afterwards, sign_unops, sign_binops and sign_quants could be merged into a single sign_pp_hints:pp_hint SymMap.t ref. In Sig_state.t, I think that unops and binops can be removed, because an identifier is identified as a symbol thanks to Sign.sign_symbols in case of a qualified identifier, or thanks to Sig_state.in_scope in case of an unqualified identifier. Then, you can use Sig_state.pp_hints to know if what is its notation.

Reply:
It doesn't even return a sym, I think that calling `find_sym` on, say, `+` which has been declared infix just raises `Not_found`. The Earley parser would add the tuple `(op, prop, qid)` (which would be `("+", 1.0, plus)` in our case) into the syntax, and scoping would only use the `qid` (so the `plus`) rather than the `op`. Here we'd like to keep a mapping from `"+"` to `plus` additionally.

Reply:
I don't understand. Currently, in master, in case of a P_Iden, scope_head calls Scope.find_qid which calls Sig_state.find_sym which returns a sym. If the ident is not qualified, find_sym looks in in_scope which from a string returns a sym. And, in case of P_UnaO and P_BinO, which disappear in your PR, scope_head calls Sig_state.find_sym directly. But Earley maps unqualified identifiers corresponding to unary or binary operators in scope to some pp_hint-like data. See Parser.get_ops. Note that, while sign_unops/binops maps strings to sym+pp_hint-like data, the sym part is ignored. It could have returned this sym field as well so that there is no need to call find_sym later.

Reply:
In case of `P_UnaO` and `P_BinO`, `scope_head` calls `Sig_state.find_sym` using the `qid` associated to the binary operator. In our case, we do not have the `qid` associated with the binary operator available anymore. So what I say is that we have to fetch the qid associated to the `"+"` identifier using the signature state. Does it make sense?

Reply:
But why do you want to make a distinction between identifiers? If you have an unqualified ident that is not a variable, you can look at in_scope to get the corresponding sym and then look at pp_hints to know if it is unary, binary, etc. You can also build a qualified ident out of this sym like in pp_qualified if you need it.

Reply:
But my point is that you won't get the symbol associated to `"+"`, it is not in `in_scope`, if I understand correctly.

Reply:
Then, this is wrong and should be fixed.